onmt/trainer.py

```diff
@@ -36,6 +37,13 @@ def build_trainer(opt, device_id, model, fields, optim, model_saver=None):
     valid_loss = onmt.utils.loss.build_loss_compute(
         model, tgt_field, opt, train=False)
 
+    scores = []
+    scoring_preparator = ScoringPreparator(fields, opt)
+    if "BLEU" in opt.scores:
+        scores.append(
+            ["BLEU", [bleu_scorer, 0, 0]]
+        )
```

Comment:
Probably cleaner (and more consistent with the rest of onmt) to define this next to `ScoringPreparator` and have a `build_scorer` function there to build and return the necessary stuff.